File: privbrowser/TrackerBlockList.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;

namespace PrivBrowser
{
    /// <summary>
    /// Loads real tracker/ad domain lists (EasyList + EasyPrivacy) instead of
    /// a small hardcoded set. On first run it downloads and caches them
    /// locally; after that it loads instantly from disk and refreshes
    /// in the background every 7 days.
    /// </summary>
    public static class TrackerBlockList
    {
        private static readonly HashSet<string> BlockedDomains = new(StringComparer.OrdinalIgnoreCase);
        private static readonly object Lock = new();

        private static readonly (string Url, string CacheFile)[] Sources = new[]
        {
            ("https://easylist.to/easylist/easylist.txt", "easylist.txt"),
            ("https://easylist.to/easylist/easyprivacy.txt", "easyprivacy.txt")
        };

        private static string CacheFolder => Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "PrivBrowser", "FilterLists");

        /// <summary>
        /// Call once at startup. Loads from local cache immediately if present
        /// (so the browser is never blocked waiting on network), then
        /// refreshes from the network in the background if the cache is
        /// missing or older than 7 days.
        /// </summary>
        public static async Task InitializeAsync()
        {
            Directory.CreateDirectory(CacheFolder);

            bool anyLoadedFromDisk = false;
            foreach (var (_, cacheFile) in Sources)
            {
                var path = Path.Combine(CacheFolder, cacheFile);
                if (File.Exists(path))
                {
                    ParseIntoBlockList(await File.ReadAllLinesAsync(path));
                    anyLoadedFromDisk = true;
                }
            }

            // Seed a small built-in fallback list so blocking works even
            // before the first successful download ever completes.
            if (!anyLoadedFromDisk)
            {
                lock (Lock)
                {
                    foreach (var d in FallbackDomains) BlockedDomains.Add(d);
                }
            }

            _ = RefreshIfStaleAsync(); // fire-and-forget background refresh
        }

        private static async Task RefreshIfStaleAsync()
        {
            using var http = new HttpClient();
            http.Timeout = TimeSpan.FromSeconds(20);

            foreach (var (url, cacheFile) in Sources)
            {
                try
                {
                    var path = Path.Combine(CacheFolder, cacheFile);
                    var isStale = !File.Exists(path) ||
                        (DateTime.UtcNow - File.GetLastWriteTimeUtc(path)) > TimeSpan.FromDays(7);
                    if (!isStale) continue;

                    var text = await http.GetStringAsync(url);
                    await File.WriteAllTextAsync(path, text);
                    ParseIntoBlockList(text.Split('\n'));
                }
                catch
                {
                    // Offline or blocked — silently keep using whatever is
                    // already loaded (cached or fallback). Not fatal.
                }
            }
        }

        /// <summary>
        /// EasyList format is mostly one rule per line. We only extract the
        /// simple "||domain.com^" domain-blocking rules, which cover the
        /// large majority of ad/tracker entries and are cheap to match.
        /// Cosmetic rules (##selector) and complex regex rules are skipped.
        /// </summary>
        private static void ParseIntoBlockList(IEnumerable<string> lines)
        {
            lock (Lock)
            {
                foreach (var raw in lines)
                {
                    var line = raw.Trim();
                    if (line.Length == 0 || line.StartsWith("!") || line.StartsWith("[")) continue;
                    if (!line.StartsWith("||")) continue;

                    var end = line.IndexOfAny(new[] { '^', '/', '$' });
                    var domain = end > 2 ? line.Substring(2, end - 2) : line.Substring(2);
                    domain = domain.Trim();
                    if (domain.Length > 0 && domain.Contains('.'))
                        BlockedDomains.Add(domain);
                }
            }
        }

        public static bool IsBlocked(string url)
        {
            if (string.IsNullOrEmpty(url)) return false;
            if (!Uri.TryCreate(url, UriKind.Absolute, out var uri)) return false;

            var host = uri.Host;
            lock (Lock)
            {
                return BlockedDomains.Contains(host) ||
                    BlockedDomains.Any(d => host.EndsWith("." + d, StringComparison.OrdinalIgnoreCase));
            }
        }

        public static int LoadedCount
        {
            get { lock (Lock) return BlockedDomains.Count; }
        }

        // Used only until the first real list finishes loading/downloading.
        private static readonly string[] FallbackDomains =
        {
            "doubleclick.net", "googlesyndication.com", "googleadservices.com",
            "google-analytics.com", "googletagmanager.com", "facebook.net",
            "connect.facebook.net", "scorecardresearch.com", "adnxs.com",
            "outbrain.com", "taboola.com", "criteo.com", "hotjar.com",
            "mixpanel.com", "segment.io", "amplitude.com", "quantserve.com",
            "adsrvr.org", "moatads.com"
        };
    }
}

File: privbrowserapk/app/src/main/java/com/andarstudio/privbrowser/TrackerBlockList.kt
package com.andarstudio.privbrowser

import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentHashMap

object TrackerBlockList {
    private val domains = ConcurrentHashMap.newKeySet<String>()

    private val sources = listOf(
        "https://easylist.to/easylist/easylist.txt" to "easylist.txt",
        "https://easylist.to/easylist/easyprivacy.txt" to "easyprivacy.txt"
    )

    private val fallbackDomains = listOf(
        "doubleclick.net", "googlesyndication.com", "googleadservices.com",
        "google-analytics.com", "googletagmanager.com", "facebook.net",
        "connect.facebook.net", "scorecardresearch.com", "adnxs.com",
        "outbrain.com", "taboola.com", "criteo.com", "hotjar.com",
        "mixpanel.com", "segment.io", "amplitude.com", "quantserve.com",
        "adsrvr.org", "moatads.com"
    )

    fun initialize(cacheDir: File) {
        cacheDir.mkdirs()
        var loadedFromDisk = false
        for ((_, fileName) in sources) {
            val file = File(cacheDir, fileName)
            if (file.exists()) {
                parseInto(file.readText())
                loadedFromDisk = true
            }
        }
        if (!loadedFromDisk) domains.addAll(fallbackDomains)

        // Refresh in the background; never block app startup on network.
        Thread { refreshIfStale(cacheDir) }.start()
    }

    private fun refreshIfStale(cacheDir: File) {
        for ((url, fileName) in sources) {
            try {
                val file = File(cacheDir, fileName)
                val isStale = !file.exists() ||
                    (System.currentTimeMillis() - file.lastModified()) > 7L * 24 * 60 * 60 * 1000
                if (!isStale) continue

                val text = fetch(url)
                file.writeText(text)
                parseInto(text)
            } catch (e: Exception) {
                // Offline is fine — keep whatever's cached or the fallback list.
            }
        }
    }

    private fun fetch(urlString: String): String {
        val conn = URL(urlString).openConnection() as HttpURLConnection
        conn.connectTimeout = 15000
        conn.readTimeout = 15000
        conn.inputStream.use { stream ->
            return stream.bufferedReader().readText()
        }
    }

    private fun parseInto(text: String) {
        for (rawLine in text.lineSequence()) {
            val line = rawLine.trim()
            if (line.isEmpty() || line.startsWith("!") || line.startsWith("[")) continue
            if (!line.startsWith("||")) continue

            var endIdx = line.length
            for (i in 2 until line.length) {
                if (line[i] == '^' || line[i] == '/' || line[i] == '$') { endIdx = i; break }
            }
            val domain = line.substring(2, endIdx).trim()
            if (domain.isNotEmpty() && domain.contains('.')) domains.add(domain)
        }
    }

    fun isBlocked(urlString: String): Boolean {
        return try {
            val host = URL(urlString).host ?: return false
            if (domains.contains(host)) return true
            domains.any { host.endsWith(".$it") }
        } catch (e: Exception) {
            false
        }
    }
}
